### PR TITLE
User/nick palmar/207/aruco pose detection

### PR DESCRIPTION
**Pull request by nico-palmar**
Tested this and looked it over... looks pretty good to me. I still need to make another PR to add in the zed_wrapper into our repo. 

Steps for testing (without the zed wrapper are outlined below):
1. Refer [here](https://www.stereolabs.com/docs/ros2/) on how to clone and build the zed2 wrapper. Specifically the 'build the package' blurb at the top. Clone and build the zed2 wrapper into your workspace. 
2. Checkout this branch. Build the uwrt_mars_rover_vision package (colcon build --packages-select uwrt_mars_rover_vision)
3. Source the overlay (the setup.bash file) and run > ros2 launch uwrt_mars_rover_vision zed_vision.launch.py

Before merging I'd be good if other people took a look at this running and tested it on their local machine. Maybe @keyonjerome can help with this. We can do it in the bay together if you want.

### CODE REVIEW

**Comment by Nick-palmar** [comment]
Nico TODO: add open CV to common upstream dependencies to pass the build tests

**Comment by keyonjerome** [comment]
@Nick-palmar is there a specific OpenCV version you are using? The arm64 build seems to automatically pull in `4.2.0`. What version did you test/write code with?

File: uwrt_mars_rover_vision/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  aruco_tracker
+	ARCHIVE DESTINATION lib
+	LIBRARY DESTINATION lib 
+	RUNTIME DESTINATION bin)
+
+# Install launch files.
+install(DIRECTORY
+  launch
+  DESTINATION share/${PROJECT_NAME}/
+)
```
Comment (keyonjerome) [inline uwrt_mars_rover_vision/CMakeLists.txt:60]: add the `if(BUILD_TEST)` from the README.md to this file, and to the package.xml
```suggestion
if (BUILD_TESTING)
   # Force generation of compile_commands.json for clang-tidy
   set(CMAKE_EXPORT_COMPILE_COMMANDS ON CACHE INTERNAL "")

   # clang-format
   find_package(ament_cmake_clang_format REQUIRED)
   ament_clang_format(
   CONFIG_FILE ${CMAKE_SOURCE_DIR}/../../.clang-format
   )

   # clang-tidy
   find_package(ament_cmake_clang_tidy REQUIRED)
   ament_clang_tidy(
   ${CMAKE_BINARY_DIR}
   CONFIG_FILE ${CMAKE_SOURCE_DIR}/../../.clang-tidy
   )

   # cppcheck
   find_package(ament_cmake_cppcheck REQUIRED)
   ament_cppcheck()

   # flake8
   find_package(ament_cmake_flake8 REQUIRED)
   ament_flake8(
   CONFIG_FILE ${CMAKE_SOURCE_DIR}/../../.flake8
   )

   # xmllint
   find_package(ament_cmake_xmllint REQUIRED)
   ament_xmllint()
endif ()
```

File: uwrt_mars_rover_vision/package.xml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  <depend>sensor_msgs</depend>
+  <depend>rclcpp_components</depend>
+  <depend>cv_bridge</depend>
+  <depend>image_transport</depend>
+  <depend>libopencv-dev</depend>
+  <depend>geometry_msgs</depend>
+
+  <test_depend>ament_lint_auto</test_depend>
+  <test_depend>ament_lint_common</test_depend>
+
```
Comment (keyonjerome) [inline uwrt_mars_rover_vision/package.xml:28]: add test dependencies from README.md

File: uwrt_mars_rover_vision/src/aruco_target_tracker.cpp [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+  {
+  RCLCPP_DEBUG_STREAM(this->get_logger(), info->distortion_model);
+  RCLCPP_DEBUG_STREAM(this->get_logger(), "fx: " << info->k[0] << " | cx: " << info->k[2] << " | fy: " << info->k[4] << " | cy: " << info->k[5]);
+  int param_num = 0; 
+  for (const double &disto_param: info->d)
+  {
+      dist_coefficients_.at<double>(0, param_num) = disto_param;
+      param_num++;
+  }
+  intrinsic_calib_matrix_ = (cv::Mat_<float>(3,3) <<
```
Comment (keyonjerome) [inline uwrt_mars_rover_vision/src/aruco_target_tracker.cpp:136]: what are these magic numbers?

**Review by keyonjerome — COMMENTED** [review]
Looks like opencv is being pulled in automatically by `rosdep` nicely for the most part, except for ARM64. Won't be picky about code style or testing; add the CMake test code provided in `README.md`, and this PR is good to go. Awesome work Nico 👍

**Comment by keyonjerome** [comment]
also add your new package to `ci.yaml`

**Comment by Nick-palmar** [comment]
> 

I think 4.7.0 dev... which is prolly why stuff is breaking

File: uwrt_mars_rover_vision/src/aruco_target_tracker.cpp (hunk 3 above)
Comment (Nick-palmar) [reply uwrt_mars_rover_vision/src/aruco_target_tracker.cpp:136]: this is just making a 3x3 matrix :) think they arent magic numbers for that reason (ie they explicitly create that)

Comment (keyonjerome) [reply uwrt_mars_rover_vision/src/aruco_target_tracker.cpp:136]: o.O

File: uwrt_mars_rover_vision/CMakeLists.txt (hunk 1 above)
Comment (Nick-palmar) [reply uwrt_mars_rover_vision/CMakeLists.txt:60]: done

File: uwrt_mars_rover_vision/package.xml (hunk 2 above)
Comment (Nick-palmar) [reply uwrt_mars_rover_vision/package.xml:33]: done

**Comment by Nick-palmar** [comment]
Before merging this.. can I find out what exactly the linter wants? I ran ament_clang_format --reformat and it reformatted my stuff but clearly that's not enough... lol. Any ideas? I think later on we need to fix this/make a better workflow (def not your fault keyon, just something to do in the off season).

**Comment by keyonjerome** [comment]
@Nick-palmar you developed this using ROS2 Foxy right, not Galactic?

**Comment by Nick-palmar** [comment]
> @Nick-palmar you developed this using ROS2 Foxy right, not Galactic?

Yea that's correct - although I believe it shouldn't make a difference

**Comment by keyonjerome** [comment]
@Nick-palmar I'm wondering if the OpenCV version provided by `ros-perception-vision-opencv` is different between Foxy and Galactic. Also, how did you install OpenCV on your system? I remember you had to something funky, can you describe it? It's weird that `ros-perception-vision-opencv` pulls in OpenCV properly for AMD64, but not for the ARM64 build. Since the Jetson is ARM64, if we don't fix this issue here I doubt it'll build on the Jetson either.

**Comment by keyonjerome** [comment]
I don't really care about the `ament-flake8` lint checks that are failing on the AMD64 build, it's ridiculous that we have to waste time autoformatting Python code in the first place instead of using something like [black](https://black.readthedocs.io/en/stable/integrations/github_actions.html) like we have `ament-clang-format --reformat` for the C++ files.

**Comment by Nick-palmar** [comment]
> I don't really care about the `ament-flake8` lint checks that are failing on the AMD64 build, it's ridiculous that we have to waste time autoformatting Python code in the first place instead of using something like [black](https://black.readthedocs.io/en/stable/integrations/github_actions.html) like we have `ament-clang-format --reformat` for the C++ files.

Yea ok sounds good - I did that already for this branch

**Comment by Nick-palmar** [comment]
> @Nick-palmar I'm wondering if the OpenCV version provided by `ros-perception-vision-opencv` is different between Foxy and Galactic. Also, how did you install OpenCV on your system? I remember you had to something funky, can you describe it? It's weird that `ros-perception-vision-opencv` pulls in OpenCV properly for AMD64, but not for the ARM64 build. Since the Jetson is ARM64, if we don't fix this issue here I doubt it'll build on the Jetson either.

Well it was more my cuda version and nvidia driver version that I had to mess around with a bunch. I forget what exactly I did for opencv so maybe it wasn't too crazy. Any thoughts on what to do?

**Pull request merged**